### PR TITLE
Probapoly init and CVecorder performance improvements

### PR DESCRIPTION
**Pull request by gamecat69**
Initial commit for probapoly.
CVecorder: Updated loading screen to show progress to avoid a perception of a hang during load. Reduced number of save-state loading retries for faster loading. Fixed bug that caused recordings to be buggy if no save state file was found.

### CODE REVIEW

File: software/contrib/probapoly.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+Outputs 1 and 4 always have a probability of 100%. Outputs 2 and 5 have a default probability of 50%. Outputs 3 and 6 have a default probability of 25%.
+
+digital_in: Clock input
+analog_in: Different mode, adjusted by setting self.ainMode as follows:
+- Mode 1: Analogue input toggles double time feature
+- Mode 2: Analogue input voltage adjusts the upper poly value
+- [default] Mode 3: Analogue input voltage adjusts the probabilities of outputs 2,3,5,6 sending gates
+
+button_1: Short press (<500ms): Reduce pattern length (using manual pattern length is ON). Long Press (>500ms): Toggle doubletime feature
+button_2: Short press (<500ms): Reduce pattern length (using manual pattern length is ON). Long Press (>500ms): Toggle Manual pattern length feature
```
Comment (awonak) [inline software/contrib/probapoly.md:25]: [optional] splitting these up on separate lines might be more readable.
```suggestion
button_1: 
    Short press (<500ms): Reduce pattern length (using manual pattern length is ON). 
    Long Press (>500ms): Toggle doubletime feature
button_2: 
    Short press (<500ms): Reduce pattern length (using manual pattern length is ON). 
    Long Press (>500ms): Toggle Manual pattern length feature
```

File: software/contrib/probapoly.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+upper: 4
+lower: 3
+automatically calculated pattern length: 12
+
+### 16:15
+upper: 16
+lower: 15
+automatically calculated pattern length: 240
+
+# Example patches
```
Comment (awonak) [inline software/contrib/probapoly.md:86]: I love the example patches section!

File: software/contrib/probapoly.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+                self.clockStep +=1
+                self.step += 1
+
+                # Reached of of pattern, or a shorter patter is now needed, reset step to 0
+                if self.step >= self.patternLength + 1:
+                    self.step = 1
+
+        @b1.handler_falling
+        def b1Pressed():
+            if ticks_diff(ticks_ms(), b1.last_pressed()) >  500 and ticks_diff(ticks_ms(), b1.last_pressed()) < 3000:
```
Comment (awonak) [inline software/contrib/probapoly.py:98]: [optional] consider making `500` and `3000` constants to improve readability (the constant name adds self documenting code) and maintainability (you'll only need to change the value in one place to update both button handlers).

**Review by awonak — APPROVED** [review]
Nice work, I'm looking forward to playing with this script!

File: software/contrib/cvecorder.py [hunk 4]
```diff
@@ -82,7 +82,7 @@ def __init__(self):
         #self.CvRecording = []  # CV recorder flags
 
         # Load CV Recordings from a previously stored state on disk or initialize if blank
-        self.showLoadingScreen()
+        #self.showLoadingScreen()
```
[the remark this answers is not in the document]
Comment (gamecat69) [reply software/contrib/cvecorder.py:85]: Done!

File: software/contrib/probapoly.py (hunk 3 above)
Comment (gamecat69) [reply software/contrib/probapoly.py:98]: Good shout. I will do this on my next commit.

File: software/contrib/probapoly.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        oled.fill_rect(rectLeftX, 6, self.upperProb1//5, 8, 1)
+        oled.rect(rectRightX, 6, rectLength, 8, 1)
+        oled.fill_rect(rectRightX, 6, self.upperProb2//5, 8, 1)
+        oled.rect(rectLeftX, 22, rectLength, 8, 1)
+        oled.fill_rect(rectLeftX, 22, self.lowerProb1//5, 8, 1)
+        oled.rect(rectRightX, 22, rectLength, 8, 1)
+        oled.fill_rect(rectRightX, 22, self.lowerProb2//5, 8, 1)
+
+        if self.doubleTimeManualOverride or self.doubleTime:
+            oled.text('!¿', 100, 22, 1)
```
Comment (mjaskula) [inline software/contrib/probapoly.py:206]: [question] Do we really have this character available? `¿`

**Review by mjaskula — APPROVED** [review]
Overall this looks good to me, though I left a bunch of questions for my own understanding.

I have not yet run this script on ,y hardware.

File: software/contrib/probapoly.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        self.upperProb2 = 25
+        self.lowerProb1 = 50
+        self.lowerProb2 = 25
+        self.doubleTime = False
+        self.doubleTimeManualOverride = False
+        self.manualPatternLength = False
+        self.patternLength = self.lcm(self.upper, self.lower)
+        self.patternLengthPrevious = self.patternLength
+        
+        # Todo: Make this mode accessible from the UI
```
[the remark this answers is not in the document]
Comment (mjaskula) [reply software/contrib/probapoly.py:63]: I opened up a project suggestion as #163 and will begin work on it as my next item.

File: software/contrib/probapoly.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+upper: 1
+lower: 4
+automatically calculated pattern length: 4
+
+### 2:3 Hemiola
+upper: 2
+lower: 3
+automatically calculated pattern length: 6
+
+### 4:3 Four over Three (Pass the god damn butter)
```
[the remark this answers is not in the document]
Comment (gamecat69) [reply software/contrib/probapoly.md:80]: "Pass the god damn butter" is a common mnemonic phrase used when learning to play 4 over 3. Google it and then try and do it by clicking your fingers or knocking on a desk - it is quite rewarding!

File: software/contrib/probapoly.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (gamecat69) [reply software/contrib/probapoly.py:63]: Nice suggestion. I think having a menu would be useful for some things that you don't want quick access to. Be aware that imposing a mandatory menu system on scripts may put some people off the EuroPi as they "don't like menu diving".

File: software/contrib/probapoly.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        @din.handler
+        def clockRising():
+            for cv in cvs:
+                cv.off()
+            #self.updateScreen()
+            self.handleClock()
+            self.clockStep +=1
+            self.step += 1
+
+            # Reached of of pattern, or a shorter patter is now needed, reset step to 0
```
[the remark this answers is not in the document]
Comment (gamecat69) [reply software/contrib/probapoly.py:78]: Good spot. Fixed!

File: software/contrib/probapoly.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+            if self.step >= self.patternLength + 1:
```
[the remark this answers is not in the document]
Comment (gamecat69) [reply software/contrib/probapoly.py:79]: agreed. updated.

File: software/contrib/probapoly.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+                self.doubleTimeManualOverride = not self.doubleTimeManualOverride
+            else:
+                # Decrement pattern length by 1
+                self.patternLength -= 1
+                # Set the pattern length to the previous value for playability
+                self.patternLengthPrevious = self.patternLength
+
+        @b2.handler_falling
+        def b1Pressed():
+            if ticks_diff(ticks_ms(), b2.last_pressed()) >  500 and ticks_diff(ticks_ms(), b2.last_pressed()) < 3000:
```
[the remark this answers is not in the document]
Comment (gamecat69) [reply software/contrib/probapoly.py:109]: The < 3000 part is to work around a bug that causes phantom button presses: #132. I have found that without adding this upper limit long button presses can occur randomly without a button being pressed.

File: software/contrib/probapoly.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (mjaskula) [reply software/contrib/probapoly.md:80]: I can't do it.

File: software/contrib/probapoly.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mjaskula) [reply software/contrib/probapoly.py:63]: Nothing mandatory is intended here. Just another tool for script author's toolboxes.

File: software/contrib/probapoly.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mjaskula) [reply software/contrib/probapoly.py:109]: Ahh, ok. Could you add that detail to the issue?

File: software/contrib/probapoly.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+        @b1.handler_falling
+        def b1Pressed():
+            if ticks_diff(ticks_ms(), b1.last_pressed()) >  500 and ticks_diff(ticks_ms(), b1.last_pressed()) < 3000:
+                # toggle double-time feature
+                self.doubleTimeManualOverride = not self.doubleTimeManualOverride
+            else:
+                # Decrement pattern length by 1
+                self.patternLength -= 1
+                # Set the pattern length to the previous value for playability
+                self.patternLengthPrevious = self.patternLength
```
[the remark this answers is not in the document]
Comment (gamecat69) [reply software/contrib/probapoly.py:105]: Good points. Your comment encouraged me to neaten up all vars and operations related to this feature. It is much simpler now. Basically, I want to enable users to switch between an auto calculated pattern length and one they specify. Calling the manual pattern length "patternLengthPrevious" was too confusing!

File: software/contrib/probapoly.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+            else:
+                # Increment pattern length by 1
+                self.patternLength += 1
+                # Set the pattern length to the previous value for playability
+                self.patternLengthPrevious = self.patternLength
+
+    def handleClock(self):
+        
+        # Play upper gate
+        if self.step == 1 or (self.step-1) % self.upper == 0:
```
[the remark this answers is not in the document]
Comment (gamecat69) [reply software/contrib/probapoly.py:126]: TBH, I can't remember! I have since simplified this section.

File: software/contrib/probapoly.py (hunk 5 above)
Comment (gamecat69) [reply software/contrib/probapoly.py:206]: No we don't. I changed this to !! instead.

File: software/contrib/probapoly.py [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+            for cv in cvs:
+                cv.off()
+            if self.doubleTimeManualOverride or self.doubleTime:
+                #self.updateScreen()
+                self.handleClock()
+                self.clockStep +=1
+                self.step += 1
+
+                # Reached of of pattern, or a shorter patter is now needed, reset step to 0
+                if self.step >= self.patternLength + 1:
```
[the remark this answers is not in the document]
Comment (gamecat69) [reply software/contrib/probapoly.py]: already done

File: software/contrib/probapoly.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+                self.patternLengthPrevious = self.patternLength
+
+        @b2.handler_falling
+        def b1Pressed():
+            if ticks_diff(ticks_ms(), b2.last_pressed()) >  500 and ticks_diff(ticks_ms(), b2.last_pressed()) < 3000:
+                self.manualPatternLength = not self.manualPatternLength
+                if self.manualPatternLength:
+                    # Set the pattern length to the previous value for playability, or 32 if not set previously (un changed)
+                    if self.patternLengthPrevious == self.patternLength:
+                        self.patternLength
```
[the remark this answers is not in the document]
Comment (gamecat69) [reply software/contrib/probapoly.py]: already corrected in new commit

File: software/contrib/probapoly.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (gamecat69) [reply software/contrib/probapoly.py:107]: Done. Thanks for the suggestion.

File: software/contrib/probapoly.py (hunk 3 above)
Comment (gamecat69) [reply software/contrib/probapoly.py]: Done!

**Comment by awonak** [comment]
I just tested both scripts, LGTM!

**Pull request merged**